src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,518 @@
+        ngx_http_lua_run_posted_threads(c, vm, r, ctx, c->requests);
+        return;
+    }
+
+    /* rc == NGX_ERROR || rc >= NGX_OK */
+
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return;
+    }
+
+    return;
+
+failed_return:
+    ngx_http_lua_free_task_ctx(worker_thread_ctx->ctx);
+    ngx_http_lua_thread_task_free(worker_thread_ctx);
+    return;
+}
+
+
+static void
+ngx_http_lua_worker_thread_cleanup(void *data)
+{
+    ngx_http_lua_co_ctx_t *ctx = data;
+    ngx_http_lua_worker_thread_ctx_t *worker_thread_ctx = ctx->data;
+    worker_thread_ctx->is_abort = 1;
+}
+
+
+/* It's not easy to use pure ffi here, using the Lua C API for now. */
+static int
+ngx_http_lua_run_worker_thread(lua_State *L)
+{
+    ngx_http_request_t                 *r;
+    ngx_http_lua_ctx_t                 *ctx;
+    int                                 n_args;
+    ngx_str_t                           thread_pool_name;
+    ngx_thread_pool_t                  *thread_pool;
+    ngx_http_lua_task_ctx_t            *tctx;
+    lua_State                          *vm;
+    size_t                              len;
+    const char                         *mod_name;
+    const char                         *func_name;
+    int                                 rc;
+    const char                         *err;
+    int                                 i;
+    ngx_thread_task_t                  *task;
+    ngx_http_lua_worker_thread_ctx_t   *worker_thread_ctx;
+
+    r = ngx_http_lua_get_req(L);
+
+    if (r == NULL) {
+        return luaL_error(L, "no request found");
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+
+    if (ctx == NULL) {
+        return luaL_error(L, "no ctx found");
+    }
+
+    ngx_http_lua_check_context(L, ctx, NGX_HTTP_LUA_CONTEXT_YIELDABLE);
+
+    n_args = lua_gettop(L);
+
+    if (n_args < 3) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "expecting at least 3 arguments");
+        return 2;
+    }
+
+    thread_pool_name.data = (u_char*) lua_tolstring(L, 1,
+                                                    &thread_pool_name.len);
+    if (thread_pool_name.data == NULL) {
+        return luaL_error(L, "lua_tolstring failed");
+    }
+    thread_pool = ngx_thread_pool_get((ngx_cycle_t *) ngx_cycle,
+                                      &thread_pool_name);
+    if (thread_pool == NULL) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "thread_pool not found");
+        return 2;
+    }
+
+    /* get vm */
+    tctx = ngx_http_lua_get_task_ctx(L, r);
+    if (tctx == NULL) {
+        lua_pushboolean(L, 0);
+        lua_pushstring(L, "no available Lua vm");
+        return 2;
+    }
+    vm = tctx->vm;
+
+    /* push function from module require */
+    mod_name = lua_tolstring(L, 2, &len);
+    if (mod_name == NULL) {
+        ngx_http_lua_free_task_ctx(tctx);
+        return luaL_error(L, "lua_tolstring failed");
+    }
```

Comment:
we'd better move these code before get task ctx, so that we can avoid free task ctx here.